### ISSUE TITLE
Do we drop python/2.7 support completely from the gentoo/2020 stack ?

### ISSUE THREAD

**Issue by mboisson** [ComputeCanada/software-stack#10]
For example, the scipy-stack module, or boost, etc... should they still support python/2.7 ? or keep that just in the 2016/2018 stacks based on nix ?

**Comment by preney** [comment]
This is Gentoo Linux' policy on Python version 2 ( from https://wiki.gentoo.org/wiki/Project:Python ):

> By the end of 2019, Python 2.7 has reached end-of-life. Many projects have dropped Python 2 support already, and more are planning to do that. Gentoo is planning to continue providing its support where necessary for how long it's feasible to. However, this doesn't mean that we're going to go out of our way to provide support where upstream discontinued it.
> 
> More specifically:
> 
> * Packages that support both Python 2 and Python 3 but have no remaining Python 2 reverse dependencies will eventually have Python 2 support removed and become Python 3 only.
> * Packages that support only Python 2 and have little chances of being ported to Python 3 will be removed eventually. The exact time of removal depends on usefulness and maintenance costs of individual packages.
> * Python 2.7 interpreter will be kept as long as necessary and/or feasible. We'll continue patching it if necessary. You can use it along with virtualenv/pip to support your py2 apps for the time being.
> * PyPy(2.7) support in packages has been removed already.
> * PyPy2.7 interpreter will be kept as it is necessary to build PyPy3.

So if our policy is to lock in the config for Gentoo Prefix, then I feel it can be installed in Gentoo (assuming no other packages need v2.7 in the prefix --or it will be automatically brought in).

Clearly a new Gentoo Prefix config generated at a later date in time might not have Python v2.7 as all packages might have by then moved away from it or were removed from Gentoo (which is a rolling distribution). If the current Gentoo Portage database was kept frozen via modules/lmod, then there would be no issue with having it in now. Of course, later it might not be in at all. If later Python 2.7 was still desired, then one would have to (i) manually install it, (ii) place suitable ebuild files for Python 2.7 in our Gentoo overlay (so it could be installed via Gentoo's package manager), or (iii) Python 2.7 is provided via some other mechanism in the prefix.

My opinion, at least for 2020, Python 2.7 should be installed as this will allow using it for packages that have not yet transitioned or are still in the process of transitioning.

**Comment by mboisson** [comment]
Note that Gentoo's policy and availability is orthogonal. We're not installing python modules through Gentoo, but through EasyBuild. This is also not about installing the python/2.7 module (it is already installed), it is about supporting it within other modules.

**Comment by bartoldeman** [comment]
I believe there is still quite a bit of researcher code out there that relies on py2.7 and probably uses numpy too. It's annoying to change back to nixpkgs just to make that run.

Personally I'd like one (and only one) scipy-stack module with support for python2.7 which then necessarily has numpy v1.16.6 (for which we should provide a wheel). Then other modules may be able to use it if it's not too hard, but we should definitely treat 2.7 as a second-class citizen. The focus is on 3.8 and 3.7.

**Comment by ostueker** [comment]
Yes, I still encounter users that depend on legacy packages that require Python 2.7. 
I believe we should provide [Python 2.7.18 (released last month)](https://blog.python.org/2020/04/python-2718-last-release-of-python-2.html) but only build new wheels or bindings in cases where they are really needed.

**Comment by bartoldeman** [comment]
@preney yes what we settled on is that the Gentoo Prefix / 2020 module should be frozen in time as far as possible. Since I was basically beta-testing before it's now at a checkout from gentoo-20200504 (using webrsync) and any modifications come from the https://github.com/ComputeCanada/gentoo-overlay repo.

Since there are still various programs in there that hard-rely on py2.7 (such as subversion) it's there but it's as minimal as possible without pip virtualenv etc. The main Python in there is Python 3.7.7 that is used by Portage, and also for EasyBuild itself. Python 3.6 isn't there.

Most researcher-oriented software with multiple versions is in modules, so we'll have modules for Qt (there is no Qt in the Gentoo module any more as there is nothing in there that we need that hard-depends on it without the qt5 use flag set), Python, GCC, etc.

**Comment by mboisson** [comment]
Discussed today at RSNT meeting.

We will install a single "scipy-stack" which will support python 2.7, 3.6, 3.7, 3.8. Future versions will drop 2.7.
